Core/OrderHelper.php

```diff
@@ -401,18 +416,45 @@ public static function onBeforeOrderCreation($oPayment)
      *
      * @param string $sUserId
      *
-     * @return Address|null
+     * @return array|null
      *
      * @throws \OxidEsales\Eshop\Core\Exception\DatabaseConnectionException
      *
      * @since 1.3.0
      */
     public static function getLastOrderShippingAddress($sUserId)
     {
-        $sOrderId = self::_getLastOrderIdFromDb($sUserId);
         $oOrder = oxNew(Order::class);
-        $oOrder->load($sOrderId);
-        return $oOrder->getDelAddressInfo();
+
+        if (!$oOrder->load(self::_getLastOrderIdFromDb($sUserId))) {
+            return null;
+        };
+
+        if (!empty($oOrder->oxorder__oxdelcountryid->value)) {
```

Comment:
The two returns are identical except one parameter. Please refactor it, one solution is making a variable of what is returned and doing it with array merge.

Reply:
the first return, returns the delivery address, the second one the billing address

Reply:
But that not a good solution, if one method decides on which data to build, then builds the "same" but two returns that are to 98% equal.

Reply:
Although these 2 returns look like they are 98% similar, they are not.
1. The first one uses OXID’s magic getters to build a shipping address using the **delivery** DB address data.
2. The second one uses OXID’s magic getters to build a shipping address using the **billing** DB address data

However, we would suggest to split it into **2 functions** to make the difference more obvious (`getDeliveryShippingAddressFromOrder` & `getBillingShippingAddressFromOrder`).

Reply:
Yes please refactor.

$oOrder->oxorder__oxdelfname->value
$oOrder->oxorder__oxbillfname->value

difference is 'del' is replaced with 'bill', so the optimal solution is 3 methods (getDeliveryShippingAddressFromOrder & getBillingShippingAddressFromOrder and a generic one that gets the 'del/bill') to set the data.

Reply:
Ok, we will split it up into **multiple methods** to make it more readable. 👍 

However, we do not think that creating a generic method for fetching 'del/bill' is a good practice as it would mean to avoid using **OXID's magic DB getters** (e.g. oxorder__oxdelfname), thus making it more complex and less readable/understandable.